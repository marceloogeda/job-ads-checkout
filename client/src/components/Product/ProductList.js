// @flow
import React, { Component } from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';

// Actions
import { addProduct } from './../../actions/order';

// Components
import Button from './../Button/Button';
import Card from './../Card/Card';
import Text from './../Text/Text';

// Types
import type { State, Dispatch } from '../../types/';
import type { Orders } from '../../types/order';
import type { PricingRule } from '../../types/pricingRule';

export type ProductListProps = {
  data: Orders,
  pricingRule: PricingRule,
  dispatch: Dispatch
};

// Styles
const Price = styled.span`
  bottom: 20px;
  position: absolute;
  right: 20px;
`;

const PriceDetail = styled.span`
  font-size: 0.85rem;
`;

class ProductList extends Component<ProductListProps, State> {
  renderAdDiscount = (product, discount) => (
    <Price>
      <Text lineThrough>${product.price}</Text>
      <Text large>{` $${product.price * discount.unit - discount.value}`}</Text>
      <PriceDetail> / per ad</PriceDetail>
    </Price>
  );

  renderBatchAdDiscount = (product, discount) => (
    <Price>
      {discount.repeatEach === 1 ? (
        <span>
          <Text>
            {`Get ${discount.unit} or more from `}
            <Text lineThrough>${product.price}</Text>
            {` to `}
          </Text>
          <Text large>${product.price - discount.value}</Text>
          <PriceDetail> / per ad</PriceDetail>
        </span>
      ) : (
        <span>
          <Text>
            {`Get ${discount.unit} for `}
            <Text large>${product.price * discount.unit - discount.value}</Text>
          </Text>
          {` or `}
          <Text>${product.price}</Text>
          <PriceDetail> / per ad</PriceDetail>
        </span>
      )}
    </Price>
  );

  renderPrice(product) {
    const { pricingRule } = this.props;

    const productDiscount =
      'discounts' in pricingRule
        ? pricingRule.discounts.find(
            discount => discount.productId === product.id
          )
        : null;

    if (productDiscount && productDiscount.unit > 1) {
      return this.renderBatchAdDiscount(product, productDiscount);
    }

    if (productDiscount && productDiscount.unit === 1) {
      return this.renderAdDiscount(product, productDiscount);
    }

    return (
      <Price>
        <Text large>${product.price}</Text>
        <PriceDetail> / per ad</PriceDetail>
      </Price>
    );
  }

  render() {
    const { data, dispatch } = this.props;

    return data.map(product => {
      return (
        <Card key={product.id}>
          {this.renderPrice(product)}
          <Text.Heading>{product.name}</Text.Heading>
          <Text.Paragraph>{product.description}</Text.Paragraph>
          <Button type="primary" onClick={() => dispatch(addProduct(product))}>
            Add to order
          </Button>
        </Card>
      );
    });
  }
}

export default connect()(ProductList);
