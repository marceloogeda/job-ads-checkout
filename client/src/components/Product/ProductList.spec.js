import React from 'react';
import TestRenderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import ProductList from './ProductList';

const setup = (setupProps: {}) => {
  const store = configureStore()();
  const renderer = TestRenderer.create(
    <ProductList store={store} data={setupProps.data} pricingRule={{}} />
  );

  return {
    store,
    renderer
  };
};

describe('# Product', () => {
  it('should render correctly', () => {
    const data = [{ id: 1, name: 'Product1', description: 'Lorem...' }];
    const { renderer } = setup({ data });

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
