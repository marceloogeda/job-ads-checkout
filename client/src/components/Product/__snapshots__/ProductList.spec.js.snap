// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Product should render correctly 1`] = `
<div
  className="sc-bwzfXH ccgmOf"
>
  <span
    className="sc-EHOje epucfV"
  >
    <span
      className="sc-ifAKCX cGPonp"
    >
      $
    </span>
    <span
      className="sc-bZQynM jGRKHg"
    >
       / per ad
    </span>
  </span>
  <h1
    className="sc-htpNat efAdPk"
  >
    Product1
  </h1>
  <p
    className="sc-bxivhb cAxnzx"
  >
    Lorem...
  </p>
  <button
    className="sc-bdVaJa dTKOxN"
    color="#0d3880"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Add to order
  </button>
</div>
`;
